internal/server/repository_workerauth.go

```diff
@@ -250,6 +246,43 @@ func StoreNodeInformationTx(ctx context.Context, reader db.Reader, writer db.Wri
 		return new(types.DuplicateRecordError)
 	}
 
+	// It's possible a connection dropped during a rotate credentials response, so the control plane's stored
+	// previous and current WorkerAuth records may not match what the worker has stored.
+	// Check what the worker indicates is its previous key and fix what we have stored before inserting the new record.
+	if node.PreviousEncryptionKey != nil {
+		query := getWorkerAuthStateByKeyIdQuery
+		rows, err := reader.Query(ctx, query, []any{sql.Named("worker_key_identifier", node.PreviousEncryptionKey.KeyId)})
+		if err != nil {
+			return errors.Wrap(ctx, err, op)
+		}
+		defer rows.Close()
+
+		var state string
+		for rows.Next() {
+			if err := reader.ScanRows(ctx, rows, &state); err != nil {
+				return errors.Wrap(ctx, err, op, errors.WithMsg("scan row failed"))
+			}
+		}
+		if err := rows.Err(); err != nil {
+			return errors.Wrap(ctx, err, op)
+		}
+
+		// If it's previous... delete current, as it's incorrect and set this one to current to
+		// ensure proper state rotation on store
+		if state == previousWorkerAuthState {
+			query := deleteWorkerAuthByKeyId
+			_, err := writer.Exec(ctx, query, []any{sql.Named("worker_key_identifier", node.PreviousEncryptionKey.KeyId)})
+			if err != nil {
+				return errors.Wrap(ctx, err, op)
+			}
+			query = updateWorkerAuthStateByKeyId
+			_, err = writer.Exec(ctx, query, []any{sql.Named("worker_key_identifier", node.PreviousEncryptionKey.KeyId)})
```

Comment:
Is `writer` a transaction? I would want to make sure both of these happen in the same transaction.

Reply:
It is, this is in `StoreNodeInformationTx` which is only ever called as part of a transaction